File: src_c/eval.c
#include "eval.h"
#include "board.h"
#include "movegen.h"
#include "util.h"
#include <stdint.h>
#include <string.h>

static int PIECE_VALS[] = {
    PAWN_VALUE,  PAWN_VALUE,  KNIGHT_VALUE, KNIGHT_VALUE,
    ROOK_VALUE,  ROOK_VALUE,  BISHOP_VALUE, BISHOP_VALUE,
    QUEEN_VALUE, QUEEN_VALUE, KING_VALUE,   KING_VALUE,
};

static int MAT_VALS[] = {
    PAWN_VALUE,  -PAWN_VALUE,  KNIGHT_VALUE, -KNIGHT_VALUE,
    ROOK_VALUE,  -ROOK_VALUE,  BISHOP_VALUE, -BISHOP_VALUE,
    QUEEN_VALUE, -QUEEN_VALUE, KING_VALUE,   -KING_VALUE,
};

static int PIECE_PHASE_VAL[] = {0, 0, 1, 1, 1, 1, 2, 2, 4, 4, 0, 0};

static int WPAWN_MID_PST[64] = {
    0,   0,   0,  0,   0,  0,   0,  0,   -35, -1, -20, -23, -15, 24, 38, -22,
    -26, -4,  -4, -10, 3,  3,   33, -12, -27, -2, -5,  12,  17,  6,  10, -25,
    -14, 13,  6,  21,  23, 12,  17, -23, -6,  7,  26,  31,  65,  56, 25, -20,
    98,  134, 61, 95,  68, 126, 34, -11, 0,   0,  0,   0,   0,   0,  0,  0,
};

static int WPAWN_END_PST[64] = {
    0,   0,   0,   0,   0,   0,   0,   0,   13, 8,   8,  10, 13, 0,  2,  -7,
    4,   7,   -6,  1,   0,   -5,  -1,  -8,  13, 9,   -3, -7, -7, -8, 3,  -1,
    32,  24,  13,  5,   -2,  4,   17,  17,  94, 100, 85, 67, 56, 53, 82, 84,
    178, 173, 158, 134, 147, 132, 165, 187, 0,  0,   0,  0,  0,  0,  0,  0,
};

static int WKNIGHT_MID_PST[64] = {
    -105, -21, -58, -33, -17,  -28, -19, -23, -29, -53, -12, -3,   -1,
    18,   -14, -19, -23, -9,   12,  10,  19,  17,  25,  -16, -13,  4,
    16,   13,  28,  19,  21,   -8,  -9,  17,  19,  53,  37,  69,   18,
    22,   -47, 60,  37,  65,   84,  129, 73,  44,  -73, -41, 72,   36,
    23,   62,  7,   -17, -167, -89, -34, -49, 61,  -97, -15, -107,
};

static int WKNIGHT_END_PST[64] = {
    -29, -51, -23, -15, -22, -18, -50, -64, -42, -20, -10, -5,  -2,
    -20, -23, -44, -23, -3,  -1,  15,  10,  -3,  -20, -22, -18, -6,
    16,  25,  16,  17,  4,   -18, -17, 3,   22,  22,  22,  11,  8,
    -18, -24, -20, 10,  9,   -1,  -9,  -19, -41, -25, -8,  -25, -2,
    -9,  -25, -24, -52, -58, -38, -13, -28, -31, -27, -63, -99,
};

static int WBISHOP_MID_PST[64] = {
    -33, -3, -14, -21, -13, -12, -39, -21, 4,   15, 16,  0,   7,   21,  33, 1,
    0,   15, 15,  15,  14,  27,  18,  10,  -6,  13, 13,  26,  34,  12,  10, 4,
    -4,  5,  19,  50,  37,  37,  7,   -2,  -16, 37, 43,  40,  35,  50,  37, -2,
    -26, 16, -18, -13, 30,  59,  18,  -47, -29, 4,  -82, -37, -25, -42, 7,  -8,
};

static int WBISHOP_END_PST[64] = {
    -23, -9, -23, -5,  -9, -16, -5, -17, -14, -18, -7,  -1, 4,  -9, -15, -27,
    -12, -3, 8,   10,  13, 3,   -7, -15, -6,  3,   13,  19, 7,  10, -3,  -9,
    -3,  9,  12,  9,   14, 10,  3,  2,   2,   -8,  0,   -1, -2, 6,  0,   4,
    -8,  -4, 7,   -12, -3, -13, -4, -14, -14, -21, -11, -8, -7, -9, -17, -24,
};

static int WROOK_MID_PST[64] = {
    -19, -13, 1,   17,  16, 7,  -37, -26, -44, -16, -20, -9, -1, 11, -6, -71,
    -45, -25, -16, -17, 3,  0,  -5,  -33, -36, -26, -12, -1, 9,  -7, 6,  -23,
    -24, -11, 7,   26,  24, 35, -8,  -20, -5,  19,  26,  36, 17, 45, 61, 16,
    27,  32,  58,  62,  80, 67, 26,  44,  32,  42,  32,  51, 63, 9,  31, 43,
};

static int WROOK_END_PST[64] = {
    -9, 2,  3,  -1, -5, -13, 4,  -20, -6, -6, 0,  2,  -9, -9, -11, -3,
    -4, 0,  -5, -1, -7, -12, -8, -16, 3,  5,  8,  4,  -5, -6, -8,  -11,
    4,  3,  13, 1,  2,  1,   -1, 2,   7,  7,  7,  5,  4,  -3, -5,  -3,
    11, 13, 13, 11, -3, 3,   8,  3,   13, 10, 18, 15, 12, 12, 8,   5,
};

static int WQUEEN_MID_PST[64] = {
    -1,  -18, -9,  10,  -15, -25, -31, -50, -35, -8,  11, 2,   8,  15, -3, 1,
    -14, 2,   -11, -2,  -5,  2,   14,  5,   -9,  -26, -9, -10, -2, -4, 3,  -3,
    -27, -27, -16, -16, -1,  17,  -2,  1,   -13, -17, 7,  8,   29, 56, 47, 57,
    -24, -39, -5,  1,   -16, 57,  28,  54,  -28, 0,   29, 12,  59, 44, 43, 45,
};

static int WQUEEN_END_PST[64] = {
    -33, -28, -22, -43, -5,  -32, -20, -41, -22, -23, -30, -16, -16,
    -23, -36, -32, -16, -27, 15,  6,   9,   17,  10,  5,   -18, 28,
    19,  47,  31,  34,  39,  23,  3,   22,  24,  45,  57,  40,  57,
    36,  -20, 6,   9,   49,  47,  35,  19,  9,   -17, 20,  32,  41,
    58,  25,  30,  0,   -9,  22,  22,  27,  27,  19,  10,  20,
};

static int WKING_MID_PST[64] = {
    -15, 36,  12,  -54, 8,   -28, 24,  14,  1,   7,   -8,  -64, -43,
    -16, 9,   8,   -14, -14, -22, -46, -44, -30, -15, -27, -49, -1,
    -27, -39, -46, -44, -33, -51, -17, -20, -12, -27, -30, -25, -14,
    -36, -9,  24,  2,   -16, -20, 6,   22,  -22, 29,  -1,  -20, -7,
    -8,  -4,  -38, -29, -65, 23,  16,  -15, -56, -34, 2,   13,
};

static int WKING_END_PST[64] = {
    -53, -34, -21, -11, -28, -14, -24, -43, -27, -11, 4,  13,  14,
    4,   -5,  -17, -19, -3,  11,  21,  23,  16,  7,   -9, -18, -4,
    21,  24,  27,  23,  9,   -11, -8,  22,  24,  27,  26, 33,  26,
    3,   10,  17,  23,  15,  20,  45,  44,  13,  -12, 17, 14,  17,
    17,  38,  23,  11,  -74, -35, -18, -18, -11, 15,  4,  -17,
};

static int BPAWN_MID_PST[64];
static int BPAWN_END_PST[64];
static int BKNIGHT_MID_PST[64];
static int BKNIGHT_END_PST[64];
static int BBISHOP_MID_PST[64];
static int BBISHOP_END_PST[64];
static int BROOK_MID_PST[64];
static int BROOK_END_PST[64];
static int BQUEEN_MID_PST[64];
static int BQUEEN_END_PST[64];
static int BKING_MID_PST[64];
static int BKING_END_PST[64];

static void flip_pst(const int *restrict src, int *restrict dest) {
  for (int i = 0; i < 64; i++) {
    int rank = i / 8, file = i % 8;
    int idx = (7 - rank) * 8 + file;
    dest[idx] = -src[i];
  }
}

static int MID_PST[12][64];
static int END_PST[12][64];

void init_eval(void) {
  flip_pst(WPAWN_MID_PST, BPAWN_MID_PST);
  flip_pst(WPAWN_END_PST, BPAWN_END_PST);
  flip_pst(WKNIGHT_MID_PST, BKNIGHT_MID_PST);
  flip_pst(WKNIGHT_END_PST, BKNIGHT_END_PST);
  flip_pst(WBISHOP_MID_PST, BBISHOP_MID_PST);
  flip_pst(WBISHOP_END_PST, BBISHOP_END_PST);
  flip_pst(WROOK_MID_PST, BROOK_MID_PST);
  flip_pst(WROOK_END_PST, BROOK_END_PST);
  flip_pst(WQUEEN_MID_PST, BQUEEN_MID_PST);
  flip_pst(WQUEEN_END_PST, BQUEEN_END_PST);
  flip_pst(WKING_MID_PST, BKING_MID_PST);
  flip_pst(WKING_END_PST, BKING_END_PST);

  memcpy(MID_PST[0], WPAWN_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[0], WPAWN_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[2], WKNIGHT_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[2], WKNIGHT_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[4], WBISHOP_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[4], WBISHOP_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[6], WROOK_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[6], WROOK_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[8], WQUEEN_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[8], WQUEEN_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[10], WKING_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[10], WKING_END_PST, 64 * sizeof(int));

  memcpy(MID_PST[1], BPAWN_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[1], BPAWN_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[3], BKNIGHT_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[3], BKNIGHT_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[5], BBISHOP_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[5], BBISHOP_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[7], BROOK_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[7], BROOK_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[9], BQUEEN_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[9], BQUEEN_END_PST, 64 * sizeof(int));
  memcpy(MID_PST[11], BKING_MID_PST, 64 * sizeof(int));
  memcpy(END_PST[11], BKING_END_PST, 64 * sizeof(int));
}

int mat_val(enum Piece p) { return MAT_VALS[p]; }
int piece_val(enum Piece p) { return PIECE_VALS[p]; }
int mg_pst_val(enum Piece p, int sq) { return MID_PST[p][sq]; }
int eg_pst_val(enum Piece p, int sq) { return END_PST[p][sq]; }

void eval_board(struct Board *b, int *mg, int *eg) {
  *mg = 0;
  *eg = 0;

  for (enum Piece p = PAWN_W; p <= KING_B; p++) {
    BB bb = b->pieces[p];

    while (bb) {
      int sq = lsb_idx(bb);
      bb &= bb - 1;

      *mg += MID_PST[p][sq];
      *mg += MAT_VALS[p];
      *eg += END_PST[p][sq];
      *eg += MAT_VALS[p];
    }
  }
}

int eval_position(struct Board *b) { return b->mg_val * ((b->ctm) ? -1 : 1); }

bool is_endgame(struct Board *b) {
  return __builtin_popcount(b->util[ALL] ^ b->pieces[PAWN_W] ^
                            b->pieces[PAWN_B]) < 5;
}

/* static int static_exchange_eval(struct Board *b, int sq, enum Piece att, */
/*                                 enum Piece vic) { */
/*   int score = PIECE_VALS[vic]; */

/*   vic = att; */
/*   while () */
/* } */

static int score_cap(Move m) {
  enum Piece vic = m_xpiece(m);
  enum Piece att = m_piece(m);

  switch (m_move_type(m)) {
  case CAP: {
    float val = (float)PIECE_VALS[vic] / (float)PIECE_VALS[att] * MVVLVA_MUL;
    return (int)val;
  }
  case EP: {
    float val =
        (float)PIECE_VALS[PAWN_W] / (float)PIECE_VALS[PAWN_W] * MVVLVA_MUL;
    return (int)val;
  }
  case NPROMOCAP:
  case RPROMOCAP:
  case BPROMOCAP:
    return -10; // undervalue underpromos as they are *probably* not useful
  case QPROMOCAP: {
    float val =
        (float)PIECE_VALS[vic] / (float)PIECE_VALS[QUEEN_W] * MVVLVA_MUL;
    return (int)val;
  }
  default:
    return 0;
  }
}

int score_quiet(Move m) {
  enum Piece p = m_piece(m);
  return PIECE_VALS[p];
}

int eval_move(Move m) {
  int val = m_is_cap(m) ? score_cap(m) : score_quiet(m);
  LOGMOVE(m);
  LOG("m score is %d\n", val);
}

File: src/eval.zig
const std = @import("std");

const board = @import("board.zig");
const Board = board.Board;
const Piece = board.Piece;
const BB = board.BB;
const Colour = board.Colour;
const movegen = @import("movegen.zig");
const Move = movegen.Move;
const consts = @import("consts");
const util = @import("util.zig");

pub const INF: i32 = 1000000;
pub const CHECKMATE: i32 = 100000;
pub const STALEMATE: i32 = 0;

pub const PAWN_VALUE: i32 = 100;
const KNIGHT_VALUE: i32 = 325;
const ROOK_VALUE: i32 = 500;
const BISHOP_VALUE: i32 = 325;
pub const QUEEN_VALUE: i32 = 1000;
const KING_VALUE: i32 = 99999;

pub const PIECE_VALS: [12]i32 = .{
    PAWN_VALUE,
    PAWN_VALUE,
    KNIGHT_VALUE,
    KNIGHT_VALUE,
    ROOK_VALUE,
    ROOK_VALUE,
    BISHOP_VALUE,
    BISHOP_VALUE,
    QUEEN_VALUE,
    QUEEN_VALUE,
    KING_VALUE,
    KING_VALUE,
};

pub const MAT_SCORES: [12]i32 = .{
    PAWN_VALUE,
    -PAWN_VALUE,
    KNIGHT_VALUE,
    -KNIGHT_VALUE,
    ROOK_VALUE,
    -ROOK_VALUE,
    BISHOP_VALUE,
    -BISHOP_VALUE,
    QUEEN_VALUE,
    -QUEEN_VALUE,
    KING_VALUE,
    -KING_VALUE,
};

pub const MID_PST: [12][64]i16 = .{
    consts.WPAWN_MID_PST,
    consts.BPAWN_MID_PST,
    consts.WKNIGHT_MID_PST,
    consts.BKNIGHT_MID_PST,
    consts.WBISHOP_MID_PST,
    consts.BBISHOP_MID_PST,
    consts.WROOK_MID_PST,
    consts.BROOK_MID_PST,
    consts.WQUEEN_MID_PST,
    consts.BQUEEN_MID_PST,
    consts.WKING_MID_PST,
    consts.BKING_MID_PST,
};

pub const END_PST: [12][64]i16 = .{
    consts.WPAWN_END_PST,
    consts.BPAWN_END_PST,
    consts.WKNIGHT_END_PST,
    consts.BKNIGHT_END_PST,
    consts.WBISHOP_END_PST,
    consts.BBISHOP_END_PST,
    consts.WROOK_END_PST,
    consts.BROOK_END_PST,
    consts.WQUEEN_END_PST,
    consts.BQUEEN_END_PST,
    consts.WKING_END_PST,
    consts.BKING_END_PST,
};

// _strongly_ inspired by PeSTO's eval func
// https://www.chessprogramming.org/PeSTO%27s_Evaluation_Function

// the value that a piece has to influence mid game / end game phase
pub const PIECE_PHASE_VAL: [12]u8 = .{ 0, 0, 1, 1, 1, 1, 2, 2, 4, 4, 0, 0 };

fn tapered_eval(mg: i32, eg: i32, phase: u8) i32 {
    const mg_phase = @min(phase, 24);
    const eg_phase = 24 - mg_phase;
    return @divTrunc(mg * mg_phase + eg * eg_phase, 24);
}

pub fn eval(b: *const Board) i32 {
    const mul: i32 = if (b.ctm == .WHITE) 1 else -1;
    return tapered_eval(b.mg_val, b.eg_val, b.phase) * mul;
}

pub fn eval_board_full(b: *const Board) struct { i32, i32, u8 } {
    var mg_val: i32 = 0;
    var eg_val: i32 = 0;
    var phase: u8 = 0;
    for (0..64) |sq| {
        const p = b.get_piece(sq);
        if (p == .NONE) continue;
        mg_val += MAT_SCORES[@intFromEnum(p)];
        mg_val += MID_PST[@intFromEnum(p)][sq];
        eg_val += MAT_SCORES[@intFromEnum(p)];
        eg_val += END_PST[@intFromEnum(p)][sq];
        phase += PIECE_PHASE_VAL[@intFromEnum(p)];
    }

    return .{ mg_val, eg_val, phase };
}

pub fn board_score(b: *const Board) i32 {
    const mul: i32 = if (b.ctm == .WHITE) 1 else -1;
    const mg_val, const eg_val, const phase = eval_board_full(b);
    return tapered_eval(mg_val, eg_val, phase) * mul;
}

const PROMO_MOVE_SCORE = 5000;
pub const CAP_MOVE_SCORE = 10000;
const TT_BEST_SCORE = 1000000;

fn least_valuable_attacker(b: *const Board, attackers: BB, ctm: Colour) struct { BB, Piece } {
    _ = ctm;
    for (Piece.pieces()) |p| {
        const pieces = attackers & b.pieces[@intFromEnum(p)];
        if (pieces > 0) return .{ board.square(@ctz(pieces)), p };
    }

    return .{ 0, .NONE };
}

// see the https://www.chessprogramming.org/SEE_-_The_Swap_Algorithm
pub fn see(b: *const Board, from: usize, to: usize, piece: Piece, xpiece: Piece) i32 {
    var p = piece;
    var from_sq = board.square(from);
    var occ = b.all_bb();

    var attackers = b.attackers_of_sq(to, .WHITE);
    attackers |= b.attackers_of_sq(to, .BLACK);

    var rook_likes: BB = 0;
    inline for ([4]Piece{ .ROOK, .ROOK_B, .QUEEN, .QUEEN_B }) |rook_like| {
        rook_likes |= b.pieces[@intFromEnum(rook_like)];
    }

    var bishop_likes: BB = 0;
    inline for ([4]Piece{ .BISHOP, .BISHOP_B, .QUEEN, .QUEEN_B }) |bishop_like| {
        bishop_likes |= b.pieces[@intFromEnum(bishop_like)];
    }

    // TODO could move scores[1] out of the loop as well maybe? also
    // to return before attacker init early if no attackers
    var scores: [32]i32 = undefined;
    scores[0] = PIECE_VALS[@intFromEnum(xpiece)];

    var ctm = b.ctm;
    var depth: usize = 1;
    while (true) : ({
        depth += 1;
    }) {
        // "make" the move
        scores[depth] = PIECE_VALS[@intFromEnum(p)] - scores[depth - 1];
        attackers ^= from_sq;
        occ ^= from_sq;

        ctm = ctm.opp();

        // add any discovered attackers
        if (from_sq & rook_likes > 0) {
            const xray = movegen.lookup_rook_xray(occ, attackers, to);
            attackers |= (xray & occ & rook_likes);
        }

        if (from_sq & bishop_likes > 0) {
            const xray = movegen.lookup_bishop_xray(occ, attackers, to);
            attackers |= (xray & occ & bishop_likes);
        }

        // util.log_bb(attackers, std.log.debug);
        from_sq, p = least_valuable_attacker(b, attackers, ctm);

        if (p == .NONE) break;
    }

    depth -= 1;

    while (depth > 0) : (depth -= 1) {
        const score = -@max(-scores[depth - 1], scores[depth]);
        scores[depth - 1] = score;
    }

    return scores[0];
}

fn mvvlva(piece: Piece, xpiece: Piece) i32 {
    return PIECE_VALS[@intFromEnum(xpiece)] - PIECE_VALS[@intFromEnum(piece)];
}

pub fn score_move(m: Move, b: *const Board, tt_bestmove: ?Move) i32 {
    if (tt_bestmove) |bm| if (movegen.moves_eq(m, bm)) return TT_BEST_SCORE;

    return switch (m.mt) {
        .QUIET, .DOUBLE, .WKINGSIDE, .BKINGSIDE, .WQUEENSIDE, .BQUEENSIDE => PIECE_VALS[@intFromEnum(m.piece)],
        .PROMO => PROMO_MOVE_SCORE + PIECE_VALS[@intFromEnum(m.xpiece)],
        .NPROMOCAP => CAP_MOVE_SCORE + see(b, m.from, m.to, .KNIGHT, m.xpiece),
        .RPROMOCAP => CAP_MOVE_SCORE + see(b, m.from, m.to, .ROOK, m.xpiece),
        .BPROMOCAP => CAP_MOVE_SCORE + see(b, m.from, m.to, .BISHOP, m.xpiece),
        .QPROMOCAP => CAP_MOVE_SCORE + see(b, m.from, m.to, .QUEEN, m.xpiece),
        .CAP, .EP => CAP_MOVE_SCORE + see(b, m.from, m.to, m.piece, m.xpiece),
    };
}
